In [1]:
import os
import numpy as np
import pandas as pd
import copy
import matplotlib.pyplot as plt 
import matplotlib.dates as mdates

In [2]:
file_name = '/Users/yinchiahuang/Library/Mobile Documents/com~apple~CloudDocs/CS677/Assignments/yinchia@bu.edu_hw1/HMC_Honda Motor Co., Ltd.csv'
df = pd.read_csv(file_name)
df

Date       Open       High        Low      Close  Adj Close  \
0      2013/9/16  38.830002  38.990002  38.669998  38.830002  34.026104   
1      2013/9/17  38.490002  38.799999  38.360001  38.669998  33.885891   
2      2013/9/18  38.720001  39.860001  38.720001  39.799999  34.876091   
3      2013/9/19  39.770000  39.830002  39.529999  39.590000  34.692081   
4      2013/9/20  39.200001  39.340000  38.830002  38.900002  34.087433   
5      2013/9/23  39.029999  39.040001  38.700001  38.919998  34.104969   
6      2013/9/24  38.990002  39.240002  38.820000  38.959999  34.140015   
7      2013/9/25  38.869999  38.900002  38.549999  38.759998  34.135738   
8      2013/9/26  39.070000  39.270000  38.889999  39.189999  34.514431   
9      2013/9/27  38.689999  38.779999  38.599998  38.720001  34.100506   
10     2013/9/30  38.049999  38.250000  37.840000  38.139999  33.589699   
11     2013/10/1  38.400002  38.549999  38.080002  38.220001  33.660168   
12     2013/10/2  38.099998  38.279999  37.709999  38.279999  33.712997   
13     2013/10/3  38.340000  38.439999  37.919998  38.070000  33.528057   
14     2013/10/4  38.180000  38.400002  37.990002  38.369999  33.792271   
15     2013/10/7  37.919998  38.230000  37.759998  37.910000  33.387142   
16     2013/10/8  38.299999  38.349998  37.849998  37.959999  33.431179   
17     2013/10/9  38.799999  39.090000  38.500000  39.070000  34.408760   
18    2013/10/10  39.770000  40.080002  39.700001  40.080002  35.298264   
19    2013/10/11  39.779999  40.119999  39.599998  40.119999  35.333477   
20    2013/10/14  39.830002  40.330002  39.689999  40.160000  35.368710   
21    2013/10/15  39.900002  39.980000  39.570000  39.570000  34.849102   
22    2013/10/16  39.639999  39.970001  39.570000  39.950001  35.183765   
23    2013/10/17  40.040001  40.410000  39.990002  40.389999  35.571266   
24    2013/10/18  40.380001  40.490002  40.160000  40.400002  35.580086   
25    2013/10/21  40.419998  40.470001  40.240002  40.470001  35.641727   
26    2013/10/22  40.439999  40.590000  40.349998  40.540001  35.703373   
27    2013/10/23  39.380001  39.459999  39.139999  39.290001  34.602505   
28    2013/10/24  39.740002  39.880001  39.520000  39.779999  35.034039   
29    2013/10/25  39.740002  39.750000  39.480000  39.650002  34.919556   
...          ...        ...        ...        ...        ...        ...   
1230    2018/8/3  30.190001  30.660000  30.190001  30.660000  30.660000   
1231    2018/8/6  30.450001  30.469999  30.270000  30.290001  30.290001   
1232    2018/8/7  30.750000  30.790001  30.670000  30.730000  30.730000   
1233    2018/8/8  30.740000  30.770000  30.670000  30.730000  30.730000   
1234    2018/8/9  30.549999  30.549999  30.180000  30.240000  30.240000   
1235   2018/8/10  30.200001  30.240000  29.959999  30.090000  30.090000   
1236   2018/8/13  29.799999  29.990000  29.629999  29.690001  29.690001   
1237   2018/8/14  29.910000  30.020000  29.870001  29.940001  29.940001   
1238   2018/8/15  29.459999  29.480000  29.200001  29.430000  29.430000   
1239   2018/8/16  29.590000  29.879999  29.580000  29.790001  29.790001   
1240   2018/8/17  29.790001  30.059999  29.740000  30.000000  30.000000   
1241   2018/8/20  30.040001  30.260000  30.020000  30.219999  30.219999   
1242   2018/8/21  30.250000  30.410000  30.250000  30.330000  30.330000   
1243   2018/8/22  30.700001  30.700001  30.450001  30.480000  30.480000   
1244   2018/8/23  30.040001  30.049999  29.830000  29.959999  29.959999   
1245   2018/8/24  29.680000  29.840000  29.639999  29.680000  29.680000   
1246   2018/8/27  30.150000  30.469999  30.020000  30.430000  30.430000   
1247   2018/8/28  30.520000  30.639999  30.469999  30.480000  30.480000   
1248   2018/8/29  30.500000  30.629999  30.400000  30.540001  30.540001   
1249   2018/8/30  30.129999  30.160000  29.959999  30.030001  30.030001   
1250   2018/8/31  29.740000  29.830000  29.510000  29.629999  29.629999   
1251    2018/9/4  29.100000  29.13999

In [3]:
def compute_percent_df(df):
    try:
        new_df = df.copy()
        new_df.rename(columns={'Adj Close': 'Price'}, inplace = True)
        new_df = new_df[['Date', 'Price']]
        new_df['price_change(%)'] = new_df['Price'].pct_change() * 100
        new_df.fillna(0, inplace = True)
        return new_df
    except Exception as e:
        print(e)
        return None 

### (a)	take your stock (Oct  2, 2013 – Aug 31, 2018)
### (b) compute daily percent and define good/bad days according to %change  

In [4]:
df['Date'] = pd.to_datetime(df['Date']) 
from_ts = '2013-10-02'
to_ts = '2018-08-31'
 
mask = (df['Date'] >= from_ts) & (df['Date'] <= to_ts)
df = df.loc[mask]

new_df = compute_percent_df(df)
new_df['good/bad'] = np.where(new_df['price_change(%)'] >= 0,'good','bad')
new_df = new_df.reset_index()
new_df

index       Date      Price  price_change(%) good/bad
0        12 2013-10-02  33.712997         0.000000     good
1        13 2013-10-03  33.528057        -0.548572      bad
2        14 2013-10-04  33.792271         0.788039     good
3        15 2013-10-07  33.387142        -1.198881      bad
4        16 2013-10-08  33.431179         0.131898     good
5        17 2013-10-09  34.408760         2.924160     good
6        18 2013-10-10  35.298264         2.585109     good
7        19 2013-10-11  35.333477         0.099758     good
8        20 2013-10-14  35.368710         0.099716     good
9        21 2013-10-15  34.849102        -1.469118      bad
10       22 2013-10-16  35.183765         0.960320     good
11       23 2013-10-17  35.571266         1.101363     good
12       24 2013-10-18  35.580086         0.024795     good
13       25 2013-10-21  35.641727         0.173246     good
14       26 2013-10-22  35.703373         0.172960     good
15       27 2013-10-23  34.602505        -3.083373      bad
16       28 2013-10-24  35.034039         1.247118     good
17       29 2013-10-25  34.919556        -0.326776      bad
18       30 2013-10-28  35.016434         0.277432     good
19       31 2013-10-29  35.421562         1.156965     good
20       32 2013-10-30  35.342289        -0.223799      bad
21       33 2013-10-31  35.192574        -0.423614      bad
22       34 2013-11-01  35.315868         0.350341     good
23       35 2013-11-04  35.289444        -0.074822      bad
24       36 2013-11-05  35.113312        -0.499107      bad
25       37 2013-11-06  35.139740         0.075265     good
26       38 2013-11-07  34.443985        -1.979966      bad
27       39 2013-11-08  34.893139         1.304013     good
28       40 2013-11-11  35.016434         0.353350     good
29       41 2013-11-12  35.289444         0.779662     good
...     ...        ...        ...              ...      ...
1209   1221 2018-07-23  29.260000         0.239805     good
1210   1222 2018-07-24  29.650000         1.332878     good
1211   1223 2018-07-25  29.860001         0.708266     good
1212   1224 2018-07-26  29.990000         0.435362     good
1213   1225 2018-07-27  30.330000         1.133711     good
1214   1226 2018-07-30  30.430000         0.329707     good
1215   1227 2018-07-31  30.660000         0.755833     good
1216   1228 2018-08-01  30.600000        -0.195695      bad
1217   1229 2018-08-02  30.110001        -1.601304      bad
1218   1230 2018-08-03  30.660000         1.826632     good
1219   1231 2018-08-06  30.290001        -1.206781      bad
1220   1232 2018-08-07  30.730000         1.452621     good
1221   1233 2018-08-08  30.730000         0.000000     good
1222   1234 2018-08-09  30.240000        -1.594533      bad
1223   1235 2018-08-10  30.090000        -0.496032      bad
1224   1236 2018-08-13  29.690001        -1.329342      bad
1225   1237 2018-08-14  29.940001         0.842034     good
1226   1238 2018-08-15  29.430000        -1.703410      bad
1227   1239 2018-08-16  29.790001         1.223245     good
1228   1240 2018-08-17  30.000000         0.704931     good
1229   1241 2018-08-20  30.219999         0.733330     good
1230   1242 2018-08-21  30.330000         0.364001     good
1231   1243 2018-08-22  30.480000         0.494560     good
1232   1244 2018-08-23  29.959999        -1.706040      bad
1233   1245 2018-08-24  29.680000        -0.934576      bad
1234   1246 2018-08-27  30.430000         2.526954     good
1235   1247 2018-08-28  30.480000         0.164312     good
1236   1248 2018-08-29  30.540001         0.196854     good
1237   1249 2018-08-30  30.030001        -1.669941      bad
1238   1250 2018-08-31  29.629999        -1.332008      bad

[1239 rows x 5 columns]

### (c)	compute total accumulation if you started with 100 dollars on October 2, 2013 assuming that you know the performance on the next day

In [5]:
def buy(cost,price):
    shares = cost / price
    return shares

def sell(cost,shares, price):
    revenue = shares * price
    profit = revenue - cost
    return profit

def profit_calculator(new_df):
    cost = 100    
    shares = []
    profit = []
    
    #new_df['buy/sell'] = np.where(new_df['price_change(%)'] >= 0,'buy','sell')
    conditions = [(new_df['good/bad']== 'good') & (new_df['good/bad'].shift(-1) == 'good'), (new_df['good/bad']== 'good') & (new_df['good/bad'].shift(-1) == 'bad'),(new_df['good/bad']== 'bad') & (new_df['good/bad'].shift(-1) == 'good'),(new_df['good/bad']== 'bad') & (new_df['good/bad'].shift(-1) == 'bad')]
    choices = ['hold','sell','buy','hold']
    new_df['buy/sell'] = np.select(conditions,choices)

    if new_df.loc[0]['buy/sell']== 'sell':
        new_df.at[0, 'buy/sell'] = 'hold'
        
    
    for row in range(len(new_df)):
        if new_df.loc[row]['buy/sell'] == 'buy':
            shares.append(buy(cost,new_df.loc[row]['Price']))    
        elif new_df.loc[row]['buy/sell'] == 'sell':
            shares.append(0)      
        else:
            if row ==0:
                shares.append(0)
            else:    
                shares.append(shares[len(shares)-1])
    new_df['Shares']= shares
    
    for row in range(len(new_df)):
        if new_df.loc[row]['buy/sell'] == 'sell':
            profit.append(sell(cost,new_df.loc[row-1]['Shares'],new_df.loc[row]['Price']))
        else:
            profit.append(0)
    
    new_df['Profit']= profit
    
    return new_df


In [6]:
profit_df = profit_calculator(new_df)

In [7]:
# count the total number of transactions
trans_df = new_df[profit_df['buy/sell']=='sell']
total_trans = len(trans_df)  
total_profit = sum(trans_df['Profit'])

print("trans = ", total_trans)
print("Profit = ", total_profit )


trans =  313
Profit =  589.2368809422805


In [8]:
profit_df

index       Date      Price  price_change(%) good/bad buy/sell  \
0        12 2013-10-02  33.712997         0.000000     good     hold   
1        13 2013-10-03  33.528057        -0.548572      bad      buy   
2        14 2013-10-04  33.792271         0.788039     good     sell   
3        15 2013-10-07  33.387142        -1.198881      bad      buy   
4        16 2013-10-08  33.431179         0.131898     good     hold   
5        17 2013-10-09  34.408760         2.924160     good     hold   
6        18 2013-10-10  35.298264         2.585109     good     hold   
7        19 2013-10-11  35.333477         0.099758     good     hold   
8        20 2013-10-14  35.368710         0.099716     good     sell   
9        21 2013-10-15  34.849102        -1.469118      bad      buy   
10       22 2013-10-16  35.183765         0.960320     good     hold   
11       23 2013-10-17  35.571266         1.101363     good     hold   
12       24 2013-10-18  35.580086         0.024795     good     hold   
13       25 2013-10-21  35.641727         0.173246     good     hold   
14       26 2013-10-22  35.703373         0.172960     good     sell   
15       27 2013-10-23  34.602505        -3.083373      bad      buy   
16       28 2013-10-24  35.034039         1.247118     good     sell   
17       29 2013-10-25  34.919556        -0.326776      bad      buy   
18       30 2013-10-28  35.016434         0.277432     good     hold   
19       31 2013-10-29  35.421562         1.156965     good     sell   
20       32 2013-10-30  35.342289        -0.223799      bad     hold   
21       33 2013-10-31  35.192574        -0.423614      bad      buy   
22       34 2013-11-01  35.315868         0.350341     good     sell   
23       35 2013-11-04  35.289444        -0.074822      bad     hold   
24       36 2013-11-05  35.113312        -0.499107      bad      buy   
25       37 2013-11-06  35.139740         0.075265     good     sell   
26       38 2013-11-07  34.443985        -1.979966      bad      buy   
27       39 2013-11-08  34.893139         1.304013     good     hold   
28       40 2013-11-11  35.016434         0.353350     good     hold   
29       41 2013-11-12  35.289444         0.779662     good     hold   
...     ...        ...        ...              ...      ...      ...   
1209   1221 2018-07-23  29.260000         0.239805     good     hold   
1210   1222 2018-07-24  29.650000         1.332878     good     hold   
1211   1223 2018-07-25  29.860001         0.708266     good     hold   
1212   1224 2018-07-26  29.990000         0.435362     good     hold   
1213   1225 2018-07-27  30.330000         1.133711     good     hold   
1214   1226 2018-07-30  30.430000         0.329707     good     hold   
1215   1227 2018-07-31  30.660000         0.755833     good     sell   
1216   1228 2018-08-01  30.600000        -0.195695      bad     hold   
1217   1229 2018-08-02  30.110001        -1.601304      bad      buy   
1218   1230 2018-08-03  30.660000         1.826632     good     sell   
1219   1231 2018-08-06  30.290001        -1.206781      bad      buy   
1220   1232 2018-08-07  30.730000         1.452621     good     hold   
1221   1233 2018-08-08  30.730000         0.000000     good     sell   
1222   1234 2018-08-09  30.240000        -1.594533      bad     hold   
1223   1235 2018-08-10  30.090000        -0.496032      bad     hold   
1224   1236 2018-08-13  29.690001        -1.329342      bad      buy   
1225   1237 2018-08-14  29.940001         0.842034     good     sell   
1226   1238 2018-08-15  29.430000        -1.703410      bad      buy   
1227   1239 2018-08-16  29.790001         1.223245     good     hold   
1228   1240 2018-08-17  30.000000         0.704931     good     hold   
1229   1241 2018-08-20  30.219999         0.733330     good     hold   
1230   1242 2018-08-21  30.330000         0.364001     good     hold   
1231   1243 2018-08-22  30.480000         0.494560     good     sell   
1232   1244 2018-08-23  29.959999        -1.70

Assuming that I know the performance on the next day, the performace would outwight the market performance. According to yahoo.finance, the 5-year trailing return rate of S&P-500 is about 14.39%. But the return rate of this case is about 589%. 

### Assume buy and hold but you missed the best 10 days

In [9]:
df['Date'] = pd.to_datetime(df['Date']) 
from_ts = '2013-10-02'
to_ts = '2018-08-31'
 
mask = (df['Date'] >= from_ts) & (df['Date'] <= to_ts)
df = df.loc[mask]

miss_good_df = compute_percent_df(df)
miss_good_df['good/bad'] = np.where(miss_good_df['price_change(%)'] >= 0,'good','bad')

/Users/yinchiahuang/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [10]:
miss_good_df= miss_good_df.sort_values(by=['price_change(%)'],ascending=False).reset_index().tail(len(miss_good_df)-10).sort_values(by=['Date'],ascending=True)
miss_good_df = miss_good_df.reset_index()

In [11]:
miss_good_df

level_0  index       Date      Price  price_change(%) good/bad
0         629     12 2013-10-02  33.712997         0.000000     good
1         883     13 2013-10-03  33.528057        -0.548572      bad
2         272     14 2013-10-04  33.792271         0.788039     good
3        1051     15 2013-10-07  33.387142        -1.198881      bad
4         559     16 2013-10-08  33.431179         0.131898     good
5          23     17 2013-10-09  34.408760         2.924160     good
6          39     18 2013-10-10  35.298264         2.585109     good
7         572     19 2013-10-11  35.333477         0.099758     good
8         573     20 2013-10-14  35.368710         0.099716     good
9        1108     21 2013-10-15  34.849102        -1.469118      bad
10        224     22 2013-10-16  35.183765         0.960320     good
11        195     23 2013-10-17  35.571266         1.101363     good
12        619     24 2013-10-18  35.580086         0.024795     good
13        539     25 2013-10-21  35.641727         0.173246     good
14        540     26 2013-10-22  35.703373         0.172960     good
15       1222     27 2013-10-23  34.602505        -3.083373      bad
16        162     28 2013-10-24  35.034039         1.247118     good
17        772     29 2013-10-25  34.919556        -0.326776      bad
18        489     30 2013-10-28  35.016434         0.277432     good
19        178     31 2013-10-29  35.421562         1.156965     good
20        737     32 2013-10-30  35.342289        -0.223799      bad
21        821     33 2013-10-31  35.192574        -0.423614      bad
22        461     34 2013-11-01  35.315868         0.350341     good
23        672     35 2013-11-04  35.289444        -0.074822      bad
24        863     36 2013-11-05  35.113312        -0.499107      bad
25        584     37 2013-11-06  35.139740         0.075265     good
26       1171     38 2013-11-07  34.443985        -1.979966      bad
27        147     39 2013-11-08  34.893139         1.304013     good
28        459     40 2013-11-11  35.016434         0.353350     good
29        275     41 2013-11-12  35.289444         0.779662     good
...       ...    ...        ...        ...              ...      ...
1199      509   1221 2018-07-23  29.260000         0.239805     good
1200      143   1222 2018-07-24  29.650000         1.332878     good
1201      304   1223 2018-07-25  29.860001         0.708266     good
1202      411   1224 2018-07-26  29.990000         0.435362     good
1203      184   1225 2018-07-27  30.330000         1.133711     good
1204      470   1226 2018-07-30  30.430000         0.329707     good
1205      281   1227 2018-07-31  30.660000         0.755833     good
1206      721   1228 2018-08-01  30.600000        -0.195695      bad
1207     1130   1229 2018-08-02  30.110001        -1.601304      bad
1208       79   1230 2018-08-03  30.660000         1.826632     good
1209     1056   1231 2018-08-06  30.290001        -1.206781      bad
1210      123   1232 2018-08-07  30.730000         1.452621     good
1211      623   1233 2018-08-08  30.730000         0.000000     good
1212     1125   1234 2018-08-09  30.240000        -1.594533      bad
1213      859   1235 2018-08-10  30.090000        -0.496032      bad
1214     1083   1236 2018-08-13  29.690001        -1.329342      bad
1215      254   1237 2018-08-14  29.940001         0.842034     good
1216     1143   1238 2018-08-15  29.430000        -1.703410      bad
1217      168   1239 2018-08-16  29.790001         1.223245     good
1218      307   1240 2018-08-17  30.000000         0.704931     good
1219      294   1241 2018-08-20  30.219999         0.733330     good
1220      448   1242 2018-08-21  30.330000         0.364001     good
1221      386   1243 2018-08-22  30.480000         0.494560     good
1222     1144   1244 2018-08-23  29.959999        -1.706040      bad
1223      990   1245 2018-08-24  29.680000        -0.934576      bad
1224       42   1246 2018-08-27  30.430000         2.526954     good
1225

In [12]:
profit_miss_good_df = profit_calculator(miss_good_df)
profit_miss_good_df

level_0  index       Date      Price  price_change(%) good/bad buy/sell  \
0         629     12 2013-10-02  33.712997         0.000000     good     hold   
1         883     13 2013-10-03  33.528057        -0.548572      bad      buy   
2         272     14 2013-10-04  33.792271         0.788039     good     sell   
3        1051     15 2013-10-07  33.387142        -1.198881      bad      buy   
4         559     16 2013-10-08  33.431179         0.131898     good     hold   
5          23     17 2013-10-09  34.408760         2.924160     good     hold   
6          39     18 2013-10-10  35.298264         2.585109     good     hold   
7         572     19 2013-10-11  35.333477         0.099758     good     hold   
8         573     20 2013-10-14  35.368710         0.099716     good     sell   
9        1108     21 2013-10-15  34.849102        -1.469118      bad      buy   
10        224     22 2013-10-16  35.183765         0.960320     good     hold   
11        195     23 2013-10-17  35.571266         1.101363     good     hold   
12        619     24 2013-10-18  35.580086         0.024795     good     hold   
13        539     25 2013-10-21  35.641727         0.173246     good     hold   
14        540     26 2013-10-22  35.703373         0.172960     good     sell   
15       1222     27 2013-10-23  34.602505        -3.083373      bad      buy   
16        162     28 2013-10-24  35.034039         1.247118     good     sell   
17        772     29 2013-10-25  34.919556        -0.326776      bad      buy   
18        489     30 2013-10-28  35.016434         0.277432     good     hold   
19        178     31 2013-10-29  35.421562         1.156965     good     sell   
20        737     32 2013-10-30  35.342289        -0.223799      bad     hold   
21        821     33 2013-10-31  35.192574        -0.423614      bad      buy   
22        461     34 2013-11-01  35.315868         0.350341     good     sell   
23        672     35 2013-11-04  35.289444        -0.074822      bad     hold   
24        863     36 2013-11-05  35.113312        -0.499107      bad      buy   
25        584     37 2013-11-06  35.139740         0.075265     good     sell   
26       1171     38 2013-11-07  34.443985        -1.979966      bad      buy   
27        147     39 2013-11-08  34.893139         1.304013     good     hold   
28        459     40 2013-11-11  35.016434         0.353350     good     hold   
29        275     41 2013-11-12  35.289444         0.779662     good     hold   
...       ...    ...        ...        ...              ...      ...      ...   
1199      509   1221 2018-07-23  29.260000         0.239805     good     hold   
1200      143   1222 2018-07-24  29.650000         1.332878     good     hold   
1201      304   1223 2018-07-25  29.860001         0.708266     good     hold   
1202      411   1224 2018-07-26  29.990000         0.435362     good     hold   
1203      184   1225 2018-07-27  30.330000         1.133711     good     hold   
1204      470   1226 2018-07-30  30.430000         0.329707     good     hold   
1205      281   1227 2018-07-31  30.660000         0.755833     good     sell   
1206      721   1228 2018-08-01  30.600000        -0.195695      bad     hold   
1207     1130   1229 2018-08-02  30.110001        -1.601304      bad      buy   
1208       79   1230 2018-08-03  30.660000         1.826632     good     sell   
1209     1056   1231 2018-08-06  30.290001        -1.206781      bad      buy   
1210      123   1232 2018-08-07  30.730000         1.452621     good     hold   
1211      623   1233 2018-08-08  30.730000         0.000000     good     sell   
1212     1125   1234 2018-08-09  30.240000        -1.594533      bad     hold   
1213      859   1235 2018-08-10  30.090000        -0.496032      bad     hold   
1214     1083   1236 2018-08-13  29.690001        -1.329342      bad      buy   
1215      254   1237 2018-08-14  29.940001         0.842034     good     sell   
1216     1143   1238 2018-08-15  29.4

In [13]:
# count the total number of transactions
miss_good_trans_df = profit_miss_good_df[profit_miss_good_df['buy/sell']=='sell']
total_trans = len(miss_good_trans_df)  
total_profit = sum(miss_good_trans_df['Profit'])

print("trans = ", total_trans)
print("Profit = ", total_profit )

trans =  308
Profit =  569.0776032434364


### Assume buy and hold but you missed the worst 10 days

In [14]:
df['Date'] = pd.to_datetime(df['Date']) 
from_ts = '2013-10-02'
to_ts = '2018-08-31'
 
mask = (df['Date'] >= from_ts) & (df['Date'] <= to_ts)
df = df.loc[mask]

miss_bad_df = compute_percent_df(df)
miss_bad_df['good/bad'] = np.where(miss_bad_df['price_change(%)'] >= 0,'good','bad')
miss_bad_df= miss_bad_df.sort_values(by=['price_change(%)'],ascending=True).reset_index().tail(len(miss_bad_df)-10).sort_values(by=['Date'],ascending=True)
miss_bad_df = miss_bad_df.reset_index()

In [15]:
miss_bad_df

level_0  index       Date      Price  price_change(%) good/bad
0         608     12 2013-10-02  33.712997         0.000000     good
1         355     13 2013-10-03  33.528057        -0.548572      bad
2         966     14 2013-10-04  33.792271         0.788039     good
3         187     15 2013-10-07  33.387142        -1.198881      bad
4         679     16 2013-10-08  33.431179         0.131898     good
5        1215     17 2013-10-09  34.408760         2.924160     good
6        1199     18 2013-10-10  35.298264         2.585109     good
7         666     19 2013-10-11  35.333477         0.099758     good
8         665     20 2013-10-14  35.368710         0.099716     good
9         130     21 2013-10-15  34.849102        -1.469118      bad
10       1014     22 2013-10-16  35.183765         0.960320     good
11       1043     23 2013-10-17  35.571266         1.101363     good
12        619     24 2013-10-18  35.580086         0.024795     good
13        699     25 2013-10-21  35.641727         0.173246     good
14        698     26 2013-10-22  35.703373         0.172960     good
15         16     27 2013-10-23  34.602505        -3.083373      bad
16       1076     28 2013-10-24  35.034039         1.247118     good
17        466     29 2013-10-25  34.919556        -0.326776      bad
18        749     30 2013-10-28  35.016434         0.277432     good
19       1060     31 2013-10-29  35.421562         1.156965     good
20        501     32 2013-10-30  35.342289        -0.223799      bad
21        417     33 2013-10-31  35.192574        -0.423614      bad
22        777     34 2013-11-01  35.315868         0.350341     good
23        566     35 2013-11-04  35.289444        -0.074822      bad
24        375     36 2013-11-05  35.113312        -0.499107      bad
25        654     37 2013-11-06  35.139740         0.075265     good
26         67     38 2013-11-07  34.443985        -1.979966      bad
27       1091     39 2013-11-08  34.893139         1.304013     good
28        779     40 2013-11-11  35.016434         0.353350     good
29        963     41 2013-11-12  35.289444         0.779662     good
...       ...    ...        ...        ...              ...      ...
1199      729   1221 2018-07-23  29.260000         0.239805     good
1200     1095   1222 2018-07-24  29.650000         1.332878     good
1201      934   1223 2018-07-25  29.860001         0.708266     good
1202      827   1224 2018-07-26  29.990000         0.435362     good
1203     1054   1225 2018-07-27  30.330000         1.133711     good
1204      768   1226 2018-07-30  30.430000         0.329707     good
1205      957   1227 2018-07-31  30.660000         0.755833     good
1206      517   1228 2018-08-01  30.600000        -0.195695      bad
1207      108   1229 2018-08-02  30.110001        -1.601304      bad
1208     1159   1230 2018-08-03  30.660000         1.826632     good
1209      182   1231 2018-08-06  30.290001        -1.206781      bad
1210     1115   1232 2018-08-07  30.730000         1.452621     good
1211      615   1233 2018-08-08  30.730000         0.000000     good
1212      113   1234 2018-08-09  30.240000        -1.594533      bad
1213      379   1235 2018-08-10  30.090000        -0.496032      bad
1214      155   1236 2018-08-13  29.690001        -1.329342      bad
1215      984   1237 2018-08-14  29.940001         0.842034     good
1216       95   1238 2018-08-15  29.430000        -1.703410      bad
1217     1070   1239 2018-08-16  29.790001         1.223245     good
1218      931   1240 2018-08-17  30.000000         0.704931     good
1219      944   1241 2018-08-20  30.219999         0.733330     good
1220      790   1242 2018-08-21  30.330000         0.364001     good
1221      852   1243 2018-08-22  30.480000         0.494560     good
1222       94   1244 2018-08-23  29.959999        -1.706040      bad
1223      248   1245 2018-08-24  29.680000        -0.934576      bad
1224     1196   1246 2018-08-27  30.430000         2.526954     good
1225

In [16]:
profit_miss_bad_df = profit_calculator(miss_bad_df)
profit_miss_bad_df

level_0  index       Date      Price  price_change(%) good/bad buy/sell  \
0         608     12 2013-10-02  33.712997         0.000000     good     hold   
1         355     13 2013-10-03  33.528057        -0.548572      bad      buy   
2         966     14 2013-10-04  33.792271         0.788039     good     sell   
3         187     15 2013-10-07  33.387142        -1.198881      bad      buy   
4         679     16 2013-10-08  33.431179         0.131898     good     hold   
5        1215     17 2013-10-09  34.408760         2.924160     good     hold   
6        1199     18 2013-10-10  35.298264         2.585109     good     hold   
7         666     19 2013-10-11  35.333477         0.099758     good     hold   
8         665     20 2013-10-14  35.368710         0.099716     good     sell   
9         130     21 2013-10-15  34.849102        -1.469118      bad      buy   
10       1014     22 2013-10-16  35.183765         0.960320     good     hold   
11       1043     23 2013-10-17  35.571266         1.101363     good     hold   
12        619     24 2013-10-18  35.580086         0.024795     good     hold   
13        699     25 2013-10-21  35.641727         0.173246     good     hold   
14        698     26 2013-10-22  35.703373         0.172960     good     sell   
15         16     27 2013-10-23  34.602505        -3.083373      bad      buy   
16       1076     28 2013-10-24  35.034039         1.247118     good     sell   
17        466     29 2013-10-25  34.919556        -0.326776      bad      buy   
18        749     30 2013-10-28  35.016434         0.277432     good     hold   
19       1060     31 2013-10-29  35.421562         1.156965     good     sell   
20        501     32 2013-10-30  35.342289        -0.223799      bad     hold   
21        417     33 2013-10-31  35.192574        -0.423614      bad      buy   
22        777     34 2013-11-01  35.315868         0.350341     good     sell   
23        566     35 2013-11-04  35.289444        -0.074822      bad     hold   
24        375     36 2013-11-05  35.113312        -0.499107      bad      buy   
25        654     37 2013-11-06  35.139740         0.075265     good     sell   
26         67     38 2013-11-07  34.443985        -1.979966      bad      buy   
27       1091     39 2013-11-08  34.893139         1.304013     good     hold   
28        779     40 2013-11-11  35.016434         0.353350     good     hold   
29        963     41 2013-11-12  35.289444         0.779662     good     hold   
...       ...    ...        ...        ...              ...      ...      ...   
1199      729   1221 2018-07-23  29.260000         0.239805     good     hold   
1200     1095   1222 2018-07-24  29.650000         1.332878     good     hold   
1201      934   1223 2018-07-25  29.860001         0.708266     good     hold   
1202      827   1224 2018-07-26  29.990000         0.435362     good     hold   
1203     1054   1225 2018-07-27  30.330000         1.133711     good     hold   
1204      768   1226 2018-07-30  30.430000         0.329707     good     hold   
1205      957   1227 2018-07-31  30.660000         0.755833     good     sell   
1206      517   1228 2018-08-01  30.600000        -0.195695      bad     hold   
1207      108   1229 2018-08-02  30.110001        -1.601304      bad      buy   
1208     1159   1230 2018-08-03  30.660000         1.826632     good     sell   
1209      182   1231 2018-08-06  30.290001        -1.206781      bad      buy   
1210     1115   1232 2018-08-07  30.730000         1.452621     good     hold   
1211      615   1233 2018-08-08  30.730000         0.000000     good     sell   
1212      113   1234 2018-08-09  30.240000        -1.594533      bad     hold   
1213      379   1235 2018-08-10  30.090000        -0.496032      bad     hold   
1214      155   1236 2018-08-13  29.690001        -1.329342      bad      buy   
1215      984   1237 2018-08-14  29.940001         0.842034     good     sell   
1216       95   1238 2018-08-15  29.4

In [17]:
# count the total number of transactions
miss_bad_trans_df = profit_miss_bad_df[profit_miss_bad_df['buy/sell']=='sell']
total_trans = len(miss_bad_trans_df)  
total_profit = sum(miss_bad_trans_df['Profit'])

print("trans = ", total_trans)
print("Profit = ", total_profit )

trans =  311
Profit =  566.3976420320155


### Conclusion
Assuming that I know the performance on the next day, the performace would outwight the market performance.<br/> 
(1) If we don't miss any good days and bad days:<br/>
According to yahoo.finance, the 5-year trailing return rate of S&P-500 is about 14.39%. But the return rate of this case is about 589%.<br/>
Total number of transactions =  313<br/> 
Total profit =  589.24<br/>
(2) If we miss the top 10 best days:<br/>
Total number of transactions =  308<br/>
Total profit =  569.08<br/>
(3) If we miss the top 10 worst days:<br/>
Total number of transactions =  311<br/>
Total profit =  566.4<br/>
